src/analyze_fs.c

```diff
@@ -351,12 +351,23 @@ static void iterate_directory(struct manifest *manifest, char *pathprefix,
 
 		string_or_die(&fullname, "%s/%s", fullpath, entry->d_name);
 		populate_file_struct(file, fullname);
-		free(fullname);
 
 		if (entry->d_type == DT_DIR) {
 			iterate_directory(manifest, pathprefix, file->filename, do_hash);
+		} else if (entry->d_type == DT_UNKNOWN) { /* fall back to lstat() */
+			struct stat sb;
```

Comment:
I didn't notice this `populate_file_struct()`. Why it's needed there then? I presumed it's called for every file in a separate thread (see `get_hash()`) to make these computations parallel. Otherwise there's no need for the `DT_DIR` optimization in the first place.